File: app/src/main/java/uk/co/ourfriendirony/medianotifier/activities/async/UpdateMediaItem.java
package uk.co.ourfriendirony.medianotifier.activities.async;

import android.content.Context;
import android.os.AsyncTask;
import android.util.Log;
import android.widget.ProgressBar;
import android.widget.Toast;

import java.lang.ref.WeakReference;

import uk.co.ourfriendirony.medianotifier.clients.Client;
import uk.co.ourfriendirony.medianotifier.db.Database;
import uk.co.ourfriendirony.medianotifier.mediaitem.MediaItem;

public class UpdateMediaItem extends AsyncTask<MediaItem, Void, String> {
    private final WeakReference<Context> context;
    private final WeakReference<ProgressBar> progressBar;
    private final Database db;
    private final Client client;

    public UpdateMediaItem(Context context, ProgressBar progressBar, Database db, Client client) {
        this.context = new WeakReference<>(context);
        this.progressBar = new WeakReference<>(progressBar);
        this.db = db;
        this.client = client;
    }

    @Override
    protected void onPreExecute() {
        super.onPreExecute();
        progressBar.get().setIndeterminate(true);
    }

    @Override
    protected String doInBackground(MediaItem... mediaItems) {
        int failed = 0;
        String result = "";

        if (mediaItems.length == 1) {
            result += db.getCoreType() + " '" + mediaItems[0].getTitle() + "' Updated";
        } else {
            result += "All " + db.getCoreType() + " Media Updated";
        }

        for (MediaItem mediaItem : mediaItems) {
            try {
                mediaItem = client.getMediaItem(mediaItem.getId());
                db.update(mediaItem);
            } catch (Exception e) {
                Log.e("[FAILED_UPDATE]", mediaItem.toString() + ": " + e.getMessage());
                failed += 1;
            }
        }

        if (failed > 0) {
            result += " [Failed=" + failed + "]";
        }

        return result;
    }

    @Override
    protected void onPostExecute(String toastMsg) {
        progressBar.get().setIndeterminate(false);
        Toast.makeText(context.get(), toastMsg, Toast.LENGTH_SHORT).show();
    }
}

File: app/src/main/kotlin/uk/co/ourfriendirony/medianotifier/activities/async/UpdateMediaItem.kt
package uk.co.ourfriendirony.medianotifier.activities.async

import android.content.Context
import android.os.Handler
import android.util.Log
import android.widget.ProgressBar
import android.widget.Toast
import uk.co.ourfriendirony.medianotifier.clients.Client
import uk.co.ourfriendirony.medianotifier.db.Database
import uk.co.ourfriendirony.medianotifier.mediaitem.MediaItem
import java.lang.ref.WeakReference

class UpdateMediaItem(
    context: Context,
    progressBar: ProgressBar?,
    db: Database?,
    client: Client?,
    private val ui: Handler,
    private vararg val mediaItems: MediaItem
) : Runnable {
    private var context: WeakReference<Context>
    private var progressBar: WeakReference<ProgressBar?>
    private var db: Database?
    private var client: Client?
//    private val ui: Handler


    init {
        this.context = WeakReference(context)
        this.progressBar = WeakReference(progressBar)
        this.db = db
        this.client = client
//        this.ui = ui
//        this.mediaItems = mediaItems
    }


    override fun run() {
        ui.post {
            progressBar.get()!!.isIndeterminate = true
        }

        var failed = 0
        var toastMsg = ""
        toastMsg += if (mediaItems.size == 1) {
            db!!.coreType + " '" + mediaItems[0].title + "' Updated"
        } else {
            "All " + db!!.coreType + " Media Updated"
        }
        for (mediaItem in mediaItems) {
            try {
                db!!.update(client!!.getMediaItem(mediaItem.id))
            } catch (e: Exception) {
                Log.e("[FAILED_UPDATE]", mediaItem.toString() + ": " + e.message)
                failed += 1
            }
        }
        if (failed > 0) {
            toastMsg += " [Failed=$failed]"
        }

        ui.post {
            progressBar.get()!!.isIndeterminate = false
            Toast.makeText(context.get(), toastMsg, Toast.LENGTH_SHORT).show()
        }
    }
}